<a href="https://colab.research.google.com/github/codedosecodes/python/blob/main/Python_mod06_DjangoFlesk.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


<img src="https://static.wixstatic.com/media/187085_663a141abb41479db38c836ad13ba2b8~mv2.png" alt="" width="200px" align="right"/>

# **Web Development**

# Learn the Basics


Authors:
[Gabriel Mauricio Cortés](https://github.com/codedosecodes/),

<style>
.alert2 {
  padding: 20px;
  background-color: #f44336;
  color: white;
  margin-bottom: 15px;
}
.alert {
  border-radius: 10px;
  background-color: #bbb;
  border: 2px solid #999;
  margin-top: 20px;
  margin-bottom: 20px;
  padding-left: 5px;
  padding-top: 10px;
  padding-bottom: 10px;
  font-family: 'Arial', sans-serif;
}

.alert > h3 {
  display: block;
  font-size: 30px;
  margin-top: 5px;
  margin-bottom: 15px;
}

.alert-info {
  background-color: #1af !important;
  border: 2px solid #04a !important;
}
div.warn {    
    background-color: #fcf2f2;
    border-color: #dFb5b4;
    border-left: 5px solid #dfb5b4;
    padding: 0.5em;
    }
</style>


<table class="table table-bordered" style="border-color:green;" align="left">
  <td>
    <a target="_blank" href="https://www.codedose.codes/"><img src="https://static.wixstatic.com/media/187085_663a141abb41479db38c836ad13ba2b8~mv2.png" width="32" height="32" />View Codedose</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/drive/1x5b8mS_pGu1mUFFyvstva2UnWoHE7hO-#scrollTo=L9Wx68gsEcNQ"><img src="https://static.wixstatic.com/media/187085_adc0ad8b754748b5b46b2ce17d6c48fb~mv2.png" width="32" height="32" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/codedosecodes/python"><img src="https://static.wixstatic.com/media/187085_a8a39d79788e40ec8c2a70d4f42ec0ec~mv2.png" width="32" height="32" />View code in GitHub</a>
  </td>
  <td>
    <a href="https://drive.google.com/file/d/1I4LgcE4WRxhLkU8_jLh7-0QEtm_4yF_D/view?usp=sharing"><img src="https://static.wixstatic.com/media/187085_dbe5b3fa6ced40fb83d52e0f38da647e~mv2.png" width="32" height="32" />Download notebook</a>
  </td>
        <td>
    <a target="_blank" href="https://nbviewer.org/github/codedosecodes/python/blob/main/Python_mod06_DjangoFlesk.ipynb"><img src="https://static.wixstatic.com/media/187085_f270d64e2bdf462c9e26477d53a86602~mv2.png" width="32" height="32" />Jupyter Notebook</a>
  </td>
    <td>
    <a target="_blank" href="https://colab.research.google.com/drive/1x5b8mS_pGu1mUFFyvstva2UnWoHE7hO-#scrollTo=L9Wx68gsEcNQ"><img src="https://static.wixstatic.com/media/187085_27f04669f65449f39f6b40b6a5945496~mv2.png" width="32" height="32" />View Video</a>
  </td>
</table>





<a ><img src="https://static.wixstatic.com/media/187085_cbb7ba6817f24ee5b35bd4b8bac14115~mv2.png" width="100%" height="100%" /></a>




Web Development: The basics of web programming with Python, including the use of frameworks such as Flask or Django, are covered.


## Objetives
### know and drive
1. Flask
2. Django



# What is Flask?

Flask is a micro web framework written in Python. It's classified as a microframework because it doesn't require specific tools or libraries. Unlike some other frameworks, Flask doesn't have a database abstraction layer, form validation, or other components. Instead, it relies on pre-existing third-party libraries for these functions. However, Flask does support extensions that can add application features as if they were native to Flask. These extensions can provide functionalities like object-relational mapping, form validation, upload handling, various open authentication technologies, and more.

Flask is a web framework that allows developers to build lightweight web applications quickly and easily with Flask Libraries. It was developed by Armin Ronacher, leader of the International Group of Python Enthusiasts(POCCO). It is basically based on the WSGI toolkit and Jinja2 templating engine.

Flask is a lightweight Python framework for web applications that provides the basics for URL routing and page rendering.

Flask is called a "micro" framework because it doesn't directly provide features like form validation, database abstraction, authentication, and so on. Such features are instead provided by special Python packages called Flask extensions. The extensions integrate seamlessly with Flask so that they appear as if they were part of Flask itself. For example, Flask doesn't provide a page template engine, but installing Flask includes the Jinja templating engine by default. For convenience, we typically speak of these defaults as part of Flask.

In [7]:
!pip install flask


## Here are some key features of Flask:

* Simplicity and Flexibility: Flask gives you the tools to build a web application but doesn't impose a specific structure. This means you have the freedom to organize your application however you want, but it also means you have to make more decisions about how to do so.

* Instance-based Development: Flask uses an application instance to handle requests and responses.

* URL Routing: Flask provides a very easy way to define routes. Routes are associations between URLs and Python functions.

* Integration with Jinja2: Flask is integrated with the Jinja2 templating engine, making it easy to create dynamic web pages with Python.

* Development and Debugging: Flask provides a development server that auto-reloads and an in-built debugger.

* Extensible: While Flask is minimalist by default, it can be extended with "extensions" that can add functionalities like database integration, user authentication, and more.

* Support for Secure Cookies: Flask can store user information using secure cookies.

* Documentation: Flask has extensive documentation and an active community, making it easy to learn and troubleshoot.#

# ngrok
ngrok is a tool that allows you to expose local servers to the public internet through secure tunnels. It's especially useful for developers who want to showcase demos, share in-progress developments, or test functionalities in local environments with external tools and services.

## Here are some key features and uses of ngrok:

* Secure Tunnels: ngrok creates a secure tunnel from the public internet (ngrok.io) to a local machine. This is useful if you're running a web server, an API, or any other type of service on your local machine and you want it to be publicly accessible.

* Development and Testing: If you're developing a web application and need to test integrations with external services (like webhooks), ngrok allows you to do so without deploying your application to a remote server.

* Live Demos: If you want to showcase a demo of your in-development application to a client or colleague without deploying it, you can use ngrok to provide a temporary public link.

* Integration with Development Tools: ngrok provides an API that allows it to be integrated with development tools and workflows.

* Web Interface: Once ngrok is running, you can access a web interface on your local machine that shows traffic details, requests, and responses, which is useful for debugging.

* Support for Multiple Protocols: While commonly used for HTTP and HTTPS, ngrok also supports tunnels for other protocols like TCP.

To use ngrok, you first need to download and run it on your local machine. Once it's running, it provides a subdomain of ngrok.io that routes traffic to your local machine.

It's important to note that while ngrok is free for basic use, it also offers paid plans that provide additional features, such as custom domains, authentication, and more simultaneous connections.

In [6]:
!pip install flask-ngrok

In [1]:
!pip install pyngrok

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 718.7/718.7 kB 9.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pyngrok: filename=pyngrok-7.0.0-py3-none-any.whl size=21129 sha256=4cab5c50a956d7565d4cf029b317d1f7e75bfdf4dec82654b3fa01fac9845756
  Stored in directory: /root/.cache/pip/wheels/60/29/7b/f64332aa7e5e88fbd56d4002185ae22dcdc83b35b3d1c2cbf5
Successfully built pyngrok


In [2]:
# !pwd
# !cd ..
# !ls
import os
os.chdir('/usr/local/lib')
!ngrok authtoken  2X8FqlyzQVhtGr5SKuG3P6IyVIZ_7dDCoGtWfCNyU1NH9vDk9
# 2XRzz0ZuGbKdChunZSwX2ZtR8nA_3vcSvLpKCGEqoBDYyBW1t


Authtoken saved to configuration file: /root/.ngrok2/ngrok.yml


The command !pip install flask-ngrok installs the flask-ngrok package in the current environment.

flask-ngrok is an extension for Flask that allows you to easily expose a Flask application running in a local environment through a secure tunnel provided by ngrok. It's especially useful in environments like Google Colab, where it's not straightforward to access the Flask server from the outside due to network restrictions.

By using flask-ngrok, you don't need to manually install and set up ngrok. The extension takes care of starting ngrok and exposing your Flask application automatically.

In [15]:
from flask import Flask
from flask_ngrok import run_with_ngrok
from pyngrok import ngrok


app = Flask(__name__)
run_with_ngrok(app)  # Initialize ngrok when the app is run

@app.route('/')

def home():
    return "<h1>This is the HOME platform to learn</h1>"

if __name__ == '__main__':
    app.run()



 * Serving Flask app '__main__'
 * Debug mode: off


INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000
INFO:werkzeug:Press CTRL+C to quit


 * Running on http://4e04-34-106-202-98.ngrok-free.app
 * Traffic stats available on http://127.0.0.1:4040


INFO:werkzeug:127.0.0.1 - - [29/Oct/2023 23:56:05] "GET / HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [29/Oct/2023 23:56:05] "GET /favicon.ico HTTP/1.1" 404 -
INFO:werkzeug:127.0.0.1 - - [29/Oct/2023 23:56:10] "GET /hello HTTP/1.1" 200 -


## CREATE ROUTES :

In the first line we create the decorator corresponding to the routes to indicate that it is a "ROUTE" of the "app" object that inherits from FLASK. We do this by adding "@app.route()" and providing as an argument the home url that corresponds to a SLASH ("/"), to indicate that it is the home page. And by the way, we must always pass it as String, that is, between quotes.
Second line.. Within the route (decorator) we create the function that will decorate this route ('/') and it must necessarily have a return !
Third line within the function we create the return, that is, what the function will return when executed. And it will be executed when the client loads the url ('/')
I hope I have been clear, after this I only have to point out two important things:

For each new URL we will have to create a new function, therefore we will have to decorate each function for each URL. So we will have several decorators of the type @app.route('url') with different functions inside for each "url" or section of our website or web application.

In [11]:
from flask import Flask
from flask_ngrok import run_with_ngrok
from pyngrok import ngrok


app = Flask(__name__)
run_with_ngrok(app)  # Initialize ngrok when the app is run

@app.route('/')

def home():
    return "<h1>This is the HOME platform to learn</h1>"

@app.route('/hello')
def hello():
    return "Hello, Flask on Google Colab!"




if __name__ == '__main__':
    app.run()




 * Serving Flask app '__main__'
 * Debug mode: off


INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000
INFO:werkzeug:Press CTRL+C to quit


 * Running on http://a988-34-106-202-98.ngrok-free.app
 * Traffic stats available on http://127.0.0.1:4040


INFO:werkzeug:127.0.0.1 - - [29/Oct/2023 23:43:27] "GET / HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [29/Oct/2023 23:43:27] "GET /favicon.ico HTTP/1.1" 404 -


### Run or start our flask application
Now we can test our application (if that's what it can be called because it doesn't do anything yet). To test our application we have a server that provides us with Flask suitable for testing, but first we must indicate in the file or in the Flask configuration which one it is. the application that should start, in what MODE and in what direction.

In [25]:
from flask import Flask
from flask_ngrok import run_with_ngrok
from pyngrok import ngrok


app = Flask(__name__)
run_with_ngrok(app)  # Initialize ngrok when the app is run

@app.route('/')

def home():
    return "<h1>This is the HOME platform to learn</h1>"

@app.route('/hello')
def hello():
    return "Hello, Flask on Google Colab!"

@app.route('/blog/<int:postID>')
def show_blog(postID):
   return 'Blog Number %d' % postID

@app.route('/rev/<float:revNo>')
def revision(revNo):
   return 'Revision Number %f' % revNo
if __name__ == '__main__':
    app.run() #Method that starts the app with the address, ports and mode



 * Serving Flask app '__main__'
 * Debug mode: off


INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000
INFO:werkzeug:Press CTRL+C to quit


 * Running on http://db0d-34-106-202-98.ngrok-free.app
 * Traffic stats available on http://127.0.0.1:4040


INFO:werkzeug:127.0.0.1 - - [30/Oct/2023 00:25:19] "GET / HTTP/1.1" 200 -
INFO:werkzeug:127.0.0.1 - - [30/Oct/2023 00:25:19] "GET /favicon.ico HTTP/1.1" 404 -
INFO:werkzeug:127.0.0.1 - - [30/Oct/2023 00:25:27] "GET /blog/11 HTTP/1.1" 200 -


Run the above code from Python Shell. Visit the URL http://localhost:5000/blog/11 in the browser.

The given number is used as argument to the show_blog() function. The browser displays the following output −

Blog Number 11
Enter this URL in the browser − http://localhost:5000/rev/1.1

The revision() function takes up the floating point number as argument. The following result appears in the browser window −

Revision Number 1.100000

In [26]:
from google.colab import files
uploaded = files.upload()

Saving algebra.pdf to algebra.pdf



# Django


Django is a high-level web development framework written in Python that encourages rapid development and clean, pragmatic design. It was designed to make it easier to build complex web applications, reducing the amount of repetitive code and speeding up the development process.

Here are some key features and concepts of Django:

* MTV (Model, Template, View): Django follows the MTV design pattern, which is a variant of the MVC (Model-View-Controller) pattern. In Django:

> *Model: Defines the database structure, including tables, relationships, and fields.
> * Template: Handles presentation and how information is displayed to the user.
> * View: Controls what is shown to the user, making decisions based on user input or a request.
* ORM (Object-Relational Mapping): Django comes with a built-in ORM that allows developers to interact with the database as if they were pure Python objects, without the need to write SQL queries.

* Automatic Admin Interface: Django includes an automatic admin interface, which is a web-based administration tool auto-generated based on the model definitions.

* Security: Django provides protection against many common web application security issues, such as cross-site scripting (XSS), cross-site request forgery (CSRF), and SQL injection.

* Friendly URLs: Django promotes the use of friendly and readable URLs, which is beneficial for both usability and SEO.

* Extensibility: Django is highly extensible, with an "apps" architecture that allows developers to break a project into multiple reusable applications.

* Active Community: Django has an active and vibrant community that contributes packages, tools, and documentation. This means there are plenty of resources available for learning and troubleshooting.

* Documentation: One of Django's strengths is its detailed and high-quality documentation.

Django is suitable for developing both small web applications and large, complex ones. It's used by many well-known organizations and websites, including Instagram, Pinterest, and The Washington Post, among others.


In [30]:
!pip install Django

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.0/8.0 MB 41.3 MB/s eta 0:00:00


How to check the version of Django

In [31]:
!pip show Django

Name: Django
Version: 4.2.6
Summary: A high-level Python web framework that encourages rapid development and clean, pragmatic design.
Home-page: https://www.djangoproject.com/
Author: Django Software Foundation
Author-email: foundation@djangoproject.com
License: BSD-3-Clause
Location: /usr/local/lib/python3.10/dist-packages
Requires: asgiref, sqlparse
Required-by: 


In [38]:
!cd portfolio/

In [39]:
!ls

algebra.pdf			  libmkl_rt.so.2		    libomptarget.sycl.wrap.so
libiomp5_db.so			  libmkl_scalapack_ilp64.so.2	    libtbbbind_2_0.so
libiomp5.so			  libmkl_scalapack_lp64.so.2	    libtbbbind_2_0.so.3
libiompstubs5.so		  libmkl_sequential.so.2	    libtbbbind_2_0.so.3.10
libmkl_avx2.so.2		  libmkl_tbb_thread.so.2	    libtbbbind_2_5.so
libmkl_avx512.so.2		  libmkl_vml_avx2.so.2		    libtbbbind_2_5.so.3
libmkl_avx.so.2			  libmkl_vml_avx512.so.2	    libtbbbind_2_5.so.3.10
libmkl_blacs_intelmpi_ilp64.so.2  libmkl_vml_avx.so.2		    libtbbbind.so
libmkl_blacs_intelmpi_lp64.so.2   libmkl_vml_cmpt.so.2		    libtbbbind.so.3
libmkl_blacs_openmpi_ilp64.so.2   libmkl_vml_def.so.2		    libtbbbind.so.3.10
libmkl_blacs_openmpi_lp64.so.2	  libmkl_vml_mc3.so.2		    libtbbmalloc_proxy.so
libmkl_cdft_core.so.2		  libmkl_vml_mc.so.2		    libtbbmalloc_proxy.so.2
libmkl_core.so.2		  libomp-device-svml.o		    libtbbmalloc_proxy.so.2.10
libmkl_def.so.2			  libomp-fallback-cstring.o	    libtbbmalloc.so
libmkl_gf_

In [32]:
!django-admin startproject portfolio

In [40]:
from google.colab.output import eval_js
print(eval_js("google.colab.kernel.proxyPort(8000)"))

https://ipj0b9qy6zr-496ff2e9c6d22116-8000-colab.googleusercontent.com/


# ***Note: References***

* https://pyformat.info/
* https://pynative.com/
* https://docs.python.org/3/
* https://www.educba.com/
* https://www.learnbyexample.org/
* https://www.programiz.com/
* http://www.btechsmartclass.com/python/Python_Tutorial_Python_Inheritance.html
* https://pynative.com/python-inheritance/
* https://www.programiz.com/python-programming/object-oriented-programming
* https://realpython.com/python3-object-oriented-programming/#class-and-instance-attributes
* https://ecomputernotes.com/python/file-handling-in-python
* https://www.askpython.com/python/python-file-handling
* https://automatetheboringstuff.com/chapter8/
* https://pynative.com/python-read-file/
* https://docs.python.org/3/library/csv.html
* https://pbpython.com/dataframe-gui-overview.html
* https://realpython.com/pandas-dataframe/
* https://data-flair.training/blogs/pandas-tutorial/
* https://realpython.com/pandas-read-write-files/
* https://www.datacamp.com/tutorial/high-performance-data-manipulation-in-python-pandas2-vs-polars
* https://www.datainsightonline.com/post/how-to-manipulate-data-with-pandas
* https://medium.com/nerd-for-tech/using-csv-data-using-pandas-part-1-138fd3c66d01
* https://www.codecademy.com/learn/data-manipulation-in-python-dvp/modules/pandas-dvp/cheatsheet
* https://www.analyticsvidhya.com/blog/2021/06/data-manipulation-using-pandas-essential-functionalities-of-pandas-you-need-to-know/
* https://www.geeksforgeeks.org/data-manipulattion-in-python-using-pandas/
* https://www.codecademy.com/learn/data-processing-pandas
* https://www.superdatascience.com/courses/data-manipulation-in-python
* https://www.geeksforgeeks.org/creating-a-pandas-dataframe/
* https://pynative.com/python-pandas-dataframe/
* https://medium.com/mlpoint/pandas-for-machine-learning-53846bc9a98b
* https://www.datacamp.com/tutorial/pandas-tutorial-dataframe-python
* https://hossainlab.github.io/pandas/notebooks/01-Getting%20Started.html
* https://class.redclara.net/halley/moncora/Jupyter_Notebooks/apiMakeSens.html
* https://realpython.com/python-sqlite-sqlalchemy/
* https://realpython.com/python-sql-libraries/
* https://en.wikiversity.org/wiki/Database_Examples/Northwind/SQLite
* https://www.sqlitetutorial.net/sqlite-sample-database/
* https://www.kaggle.com/code/alaasedeeq/chinook-sql/notebook
* https://www.datacamp.com/tutorial/sqlalchemy-tutorial-examples
* https://vegibit.com/interacting-with-a-database-using-sqlalchemy-crud-operations/
*

